packages/vats/src/lib-board.js

```diff
@@ -58,83 +177,161 @@ function makeBoard(
-      val => board.getId(val),
-      slotToVal,
-    ),
-  });
+/**
+ * @param {BoardState} state
+ * @returns {StringSlotMarshaller}
+ */
+const makeReadonlyMarshaller = state => {
+  const { valToId } = state;
+  const slotToVal = makeSlotToVal(state);
+  const valToSlot = val => {
+    if (!valToId.has(val)) {
+      // Unpublished value.
+      return null;
+    }
 
-  const board = Far('Board', {
-    getPublishingMarshaller: () => publishingMarshaller,
-    getReadonlyMarshaller: () => readonlyMarshaller,
-    // Add if not already present
-    getId: value => {
-      if (!valToId.has(value)) {
-        lastSequence += 1n;
-        const seq = lastSequence;
-
-        // Append the CRC, so that the last part of the board ID is
-        // well-distributed.
-        const crcInput = `${prefix}${seq}`;
-        const crc = calcCrc(crcInput, crcDigits);
-
-        const id = `${prefix}${crc}${seq}`;
-
-        valToId.init(value, id);
-        idToVal.init(id, value);
-      }
-      return valToId.get(value);
+    // Published value.
+    return valToId.get(val);
+  };
+  // @ts-expect-error FIXME(Endo) for an unpublished value makeMarshall will push null into slots list
+  return makeMarshal(valToSlot, slotToVal);
+};
+
+/**
+ * @param {BoardState} state
+ * @returns {StringSlotMarshaller}
+ */
+const makePublishingMarshaller = state => {
+  const slotToVal = makeSlotToVal(state);
+  // Always put the value in the board.
+  const valToSlot = val => getId(val, state);
+  return makeMarshal(valToSlot, slotToVal);
+};
+// #endregion
+
+/**
+ * A board is a two-way mapping between objects (such as issuers,
+ * brands, installation handles) and plain-data string IDs.
+ * A well-known board allows sharing objects by way of their ids.
+ *
+ * @param {import('@agoric/vat-data').Baggage} baggage for upgrade successors
+ * @see {@link packages/SwingSet/docs/vat-upgrade.md|Vat Upgrade}
+ */
+export const prepareBoardKit = baggage => {
+  return prepareExoClassKit(baggage, 'Board', BoardI, makeInit(baggage), {
+    board: {
+      /**
+       * Provide an ID for a value, generating one if not already present.
+       *
+       * Each board ID includes a CRC so that the last part is well-distributed,
+       * which mitigates typo risks.
+       *
+       * Scaling Consideration: since we cannot know when consumers
+       * are no longer interested in an ID, a board holds a strong reference
+       * to `value` for its entire lifetime. For a well-known board, this
+       * is essentially forever.
+       *
+       * @param {Key} value
+       * @throws if `value` is not a Key in the @agoric/store sense
+       */
+      getId(value) {
+        return getId(value, this.state);
+      },
+      /**
+       * @param {string} id
+       * @throws if id is not in the mapping
+       */
+      getValue(id) {
+        return getValue(id, this.state);
+      },
+      /**
+       * Adapter for NameHub lookup() protocol.
```

Comment:
Is there a URL defining NameHub lookup() protocol?

Reply:
I looked and didn't find one.

Reply:
Me neither. Maybe the comment should just be refactored to better align with e.g. https://github.com/Agoric/agoric-sdk/blob/87c304682935f39ef5e666c6cccbb560a646c804/packages/solo/CHANGELOG.md?plain=1#L94
```suggestion
       * Convenience method for recursively traversing boards to get data associated with a sequence of ids corresponding to each successive board.
       * For example, `lookup("foo", "bar")` gets the value associated with id "foo" and returns the result of treating it as a board and getting the value it associates with id "bar".
```

Reply:
something like that... though "treating it as a board" is a little off. Anything with a `lookup()` method will do.

see also https://github.com/Agoric/documentation/issues/752#issuecomment-1419707507

Reply:
Updated the suggestion accordingly.